corehq/apps/aggregate_ucrs/ingestion.py

```diff
@@ -0,0 +1,185 @@
+"""
+This module deals with data ingestion: populating the aggregate tables from other tables.
+"""
+from __future__ import absolute_import, unicode_literals
+from collections import namedtuple
+from datetime import datetime
+
+import sqlalchemy
+from sqlalchemy.dialects.postgresql import insert
+
+from corehq.apps.aggregate_ucrs.aggregations import AGG_WINDOW_START_PARAM, AGG_WINDOW_END_PARAM, \
+    TimePeriodAggregationWindow, get_time_period_class
+from corehq.apps.userreports.sql import IndicatorSqlAdapter
+
+
+AggregationParam = namedtuple('AggregationParam', 'name value mapped_column_id')
+AggregationWindow = namedtuple('AggregationWindow', 'start end')
+
+
+def populate_aggregate_table_data(aggregate_table_adapter):
+    """
+    Seeds the database table with all data from the table adapter.
+    """
+    aggregate_table_definition = aggregate_table_adapter.config
+    # get checkpoint
+    last_update = get_last_aggregate_checkpoint(aggregate_table_definition)
+    for window in get_time_aggregation_windows(aggregate_table_definition, last_update):
+        populate_aggregate_table_data_for_time_period(
+            aggregate_table_adapter, window,
+        )
+
+
+def get_last_aggregate_checkpoint(aggregate_table_definition):
+    """
+    Checkpoints indicate the last time the aggregation script successfully ran.
+    Will be used to do partial ingestion.
+    """
+    # todo
+    return None
+
+
+def get_time_aggregation_windows(aggregate_table_definition, last_update):
+    if aggregate_table_definition.time_aggregation is None:
+        # if there is no time aggregation just include a single window with no value
+        yield None
+    else:
+        start_time = get_aggregation_start_period(aggregate_table_definition, last_update)
+        end_time = get_aggregation_end_period(aggregate_table_definition, last_update)
+        assert end_time >= start_time
```

Comment:
I think this should be strictly `>` since `end_time` is not inclusive.

Reply:
so I think that if you had configured something using `date_opened` and `date_closed` on a case and had a single case that was created and closed by the same form then these would be the same. 

either way, if the start and end time fall in the same period then it doesn't matter. if the end is before the start then the aggregation will be a no-op, which is weird but I guess fine (i.e. someone just configured something weird/wrong). thinking I'll just remove the assertion.